File: src/main/kotlin/com/gemini/krakenbot/model/Asset.kt
package com.gemini.krakenbot.model

import com.fasterxml.jackson.annotation.JsonValue

@JvmInline
value class Asset(@get:JsonValue val value: String) {
    override fun toString(): String = value

    val krakenTicker: String
        get() = toKrakenTicker(value)

    val tradingPair: String
        get() = tradingPair(value)

    val isUsd: Boolean
        get() = isUsdSymbol(value)

    companion object {
        const val USD = "USD"
        const val BTC = "BTC"
        const val ETH = "ETH"
        const val DOGE = "DOGE"

        const val XBT = "XBT"
        const val XDG = "XDG"

        private val KRAKEN_TICKER_BY_SYMBOL = mapOf(
            BTC to XBT,
            DOGE to XDG
        )

        private val FALLBACK_SYMBOLS = listOf(BTC, ETH, DOGE)

        operator fun invoke(value: String): Asset = Asset(value)

        fun toKrakenTicker(symbol: String): String {
            val normalizedSymbol = normalizedSymbol(symbol)
            return KRAKEN_TICKER_BY_SYMBOL[normalizedSymbol] ?: normalizedSymbol
        }

        fun tradingPair(symbol: String): String =
            "${toKrakenTicker(symbol)}$USD"

        val BTC_USD_PAIR: String = tradingPair(BTC)

        fun fromTradingPair(pair: String, allocations: List<String>): String? {
            val normalizedPair = pair.uppercase()

            return allocations
                .filterNot(::isUsdSymbol)
                .firstOrNull { symbol -> matchesTradingPair(normalizedPair, symbol) }
                ?: allocations
                    .firstOrNull { symbol -> isUsdSymbol(symbol) && matchesTradingPair(normalizedPair, symbol) }
                ?: FALLBACK_SYMBOLS
                    .firstOrNull { symbol -> matchesTradingPair(normalizedPair, symbol) }
        }

        private fun normalizedSymbol(symbol: String): String =
            symbol.uppercase()

        private fun isUsdSymbol(symbol: String): Boolean =
            symbol.equals(USD, ignoreCase = true)

        private fun matchesTradingPair(normalizedPair: String, symbol: String): Boolean {
            val normalizedSymbol = normalizedSymbol(symbol)
            val krakenTicker = toKrakenTicker(normalizedSymbol)
            return normalizedPair.startsWith(krakenTicker) ||
                    normalizedPair.startsWith(normalizedSymbol) ||
                    normalizedPair == "${krakenTicker}USD" ||
                    normalizedPair == "${normalizedSymbol}USD" ||
                    normalizedPair == "X${krakenTicker}ZUSD" ||
                    normalizedPair == "X${normalizedSymbol}ZUSD"
        }
    }
}

File: backend/src/model/asset.ts
export class Asset {
  public readonly value: string;

  constructor(value: string) {
    this.value = value;
  }

  get isUsd(): boolean {
    return this.value.toUpperCase() === 'USD';
  }

  get krakenTicker(): string {
    return Asset.toKrakenTicker(this.value);
  }

  get tradingPair(): string {
    return Asset.tradingPair(this.value);
  }

  toString(): string {
    return this.value;
  }

  static toKrakenTicker(symbol: string): string {
    const s = symbol.toUpperCase();
    if (s === 'BTC') return 'XBT';
    if (s === 'DOGE') return 'XDG';
    return s;
  }

  static tradingPair(symbol: string): string {
    return `${Asset.toKrakenTicker(symbol)}USD`;
  }

  static readonly USD = 'USD';
  static readonly BTC = 'BTC';
  static readonly ETH = 'ETH';
  static readonly DOGE = 'DOGE';
  static readonly XBT = 'XBT';
  static readonly XDG = 'XDG';
  static readonly BTC_USD_PAIR = Asset.tradingPair(Asset.BTC);
}
